File: nodejs/azure/client/lib/azure-function-key-authenticator.js
const { EncryptionHelper } = require("@mcma/encryption");
const { AuthProvider } = require("@mcma/client");

const DefaultDecryptionKey = typeof process !== "undefined" && process.env && process.env["FunctionKeyEncryptionKey"];
const FunctionKeyHeader = "x-functions-key";

function getFunctionKey(authContext, decryptionKey) {
    if (authContext && !authContext.isEncrypted) {
        return authContext.functionKey;
    }

    if (!decryptionKey || decryptionKey.length === 0) {
        throw new Error("Function key is encrypted, but a key for decrypting it was not found in the environment variables for this application.");
    }

    return EncryptionHelper.decrypt(authContext.functionKey, decryptionKey);
}

class AzureFunctionKeyAuthenticator {
    constructor(authContext, decryptionKey) {
        decryptionKey = decryptionKey || DefaultDecryptionKey;

        let _functionKey;
        Object.defineProperties(this, {
            functionKey: {
                get: () => _functionKey || (_functionKey = getFunctionKey(authContext, decryptionKey))
            }
        });

        this.sign = (config) => {
            config.headers[FunctionKeyHeader] = this.functionKey;
        };
    }
}

AuthProvider.prototype.addAzureFunctionKeyAuth = function addAzureFunctionKeyAuth(decryptionKey) {
    return this.add("AzureFunctionKey", (authContext) => new AzureFunctionKeyAuthenticator(authContext, decryptionKey));
};

module.exports = {
    AzureFunctionKeyAuthenticator
};

File: dotnet/azure/Mcma.Azure.Client/FunctionKeys/AzureFunctionKeyAuthenticator.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using Mcma.Client;
using Mcma.Encryption;

namespace Mcma.Azure.Client.FunctionKeys
{
    public class AzureFunctionKeyAuthenticator : IAuthenticator
    {
        public AzureFunctionKeyAuthenticator(AzureFunctionKeyAuthContext authContext, string decryptionKey = null)
        {
            FunctionKey = new Lazy<string>(() => GetFunctionKey(authContext));
            DecryptionKey = decryptionKey ?? Environment.GetEnvironmentVariable(AzureConstants.FunctionKeyEncryptionKeySetting);
        }

        private Lazy<string> FunctionKey { get; }

        private string DecryptionKey { get; }

        public Task SignAsync(HttpRequestMessage request)
        {
            request.Headers.Add(AzureConstants.FunctionKeyHeader, FunctionKey.Value);
#if NET452
            return Task.FromResult(true);
#else
            return Task.CompletedTask;
#endif
        }

        private string GetFunctionKey(AzureFunctionKeyAuthContext authContext)
        {
            if (!authContext.IsEncrypted)
                return authContext.FunctionKey;

            if (string.IsNullOrWhiteSpace(DecryptionKey))
                throw new Exception($"Function key is encrypted, but a key for decrypting it was not found in the environment variables for this application.");

            return EncryptionHelper.Decrypt(authContext.FunctionKey, DecryptionKey);
        }
    }
}
